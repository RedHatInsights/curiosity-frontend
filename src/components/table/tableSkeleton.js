import React from 'react';
import PropTypes from 'prop-types';
import { TableVariant } from '@patternfly/react-table';
import { Skeleton, SkeletonSize } from '@redhat-cloud-services/frontend-components/components/cjs/Skeleton';
import Table from './table';
import { translate } from '../i18n/i18n';

/**
 * Render a table with skeleton loaders.
 *
 * @param {object} props
 * @param {string} props.className
 * @param {boolean} props.borders
 * @param {number} props.colCount
 * @param {boolean} props.isHeader
 * @param {number} props.rowCount
 * @param {Function} props.t
 * @param {string} props.variant
 * @returns {Node}
 */
const TableSkeleton = ({ className, borders, colCount, isHeader, rowCount, t, variant }) => {
  const updatedColumnHeaders = [...new Array(colCount)].map(() => <Skeleton size={SkeletonSize.md} />);

  const updatedRows = [...new Array(rowCount)].map(() => ({
    cells: [...new Array(colCount)].map(() => ({ cell: <Skeleton size={SkeletonSize.md} /> }))
  }));

  return (
    <Table
      ariaLabel={t('curiosity-inventory.tableSkeletonAriaLabel')}
      borders={borders}
      className={`curiosity-skeleton-table ${className || ''}`}
      columnHeaders={updatedColumnHeaders}
      isHeader={isHeader}
      rows={updatedRows}
      variant={variant}
    />
  );
};

/**
 * Prop types.
 *
 * @type {{borders: boolean, isHeader: boolean, colCount: number, variant: string, className: string, rowCount: number}}
 */
TableSkeleton.propTypes = {
  borders: PropTypes.bool,
  className: PropTypes.string,
  colCount: PropTypes.number,
  isHeader: PropTypes.bool,
  rowCount: PropTypes.number,
  t: PropTypes.func,
  variant: PropTypes.oneOf([...Object.values(TableVariant)])
};

/**
 * Default props.
 *
 * @type {{t: translate, borders: boolean, isHeader: boolean, colCount: number, variant: null, className: null,
 *     rowCount: number}}
 */
TableSkeleton.defaultProps = {
  borders: true,
  className: null,
  colCount: 1,
  isHeader: true,
  rowCount: 5,
  t: translate,
  variant: null
};

export { TableSkeleton as default, TableSkeleton };
