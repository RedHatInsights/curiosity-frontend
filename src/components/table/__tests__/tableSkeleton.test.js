import React from 'react';
import { shallow } from 'enzyme';
import { TableVariant } from '@patternfly/react-table';
import { TableSkeleton } from '../tableSkeleton';

describe('TableSkeleton Component', () => {
  it('should render a non-connected component', () => {
    const props = {};

    const component = shallow(<TableSkeleton {...props} />);
    expect(component).toMatchSnapshot('non-connected');
  });

  it('should allow variations in table layout', () => {
    const props = {
      colCount: 5,
      rowCount: 1
    };

    const component = shallow(<TableSkeleton {...props} />);
    expect(component).toMatchSnapshot('column and row count ');

    component.setProps({
      colWidth: [null, undefined, 30, 15, 15]
    });
    expect(component).toMatchSnapshot('table column widths');

    component.setProps({
      borders: false,
      isHeader: false
    });
    expect(component).toMatchSnapshot('borders and table header row removed');

    component.setProps({
      className: 'lorem-ipsum-class',
      variant: TableVariant.compact
    });
    expect(component).toMatchSnapshot('className and variant');

    component.setProps({
      rowCount: 0
    });
    expect(component).toMatchSnapshot('no table header and zero rows');

    component.setProps({
      isHeader: true
    });
    expect(component).toMatchSnapshot('table header and zero rows');
  });
});
