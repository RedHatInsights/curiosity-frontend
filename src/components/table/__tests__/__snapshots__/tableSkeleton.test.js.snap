// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSkeleton Component should allow variations in table layout: borders and table header row removed 1`] = `
<Table
  ariaLabel={null}
  className="curiosity-skeleton-table "
  columnHeaders={
    [
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
    ]
  }
  componentClassNames={
    {
      "table": "curiosity-table",
      "td": "curiosity-table__td",
      "tdAction": "curiosity-table__td-action",
      "tdExpand": "curiosity-table__td-expand",
      "tdExpanded": "curiosity-table__td-expand-expanded",
      "tdExpandedContent": "curiosity-table__td-expand-content",
      "tdExpandedWrapper": "curiosity-table__td-expand-wrapper",
      "tdSelect": "curiosity-table__td-select",
      "th": "curiosity-table__th",
      "tr": "curiosity-table__tr",
      "trExpand": "curiosity-table__tr-expand",
      "trExpanded": "curiosity-table__tr-expand-expanded",
      "trExpandedContent": "curiosity-table__tr-expand-content",
    }
  }
  emptyTable={
    {
      "message": "Clear all filters and try again",
      "title": "No results found",
    }
  }
  isBorders={false}
  isHeader={false}
  isStriped={false}
  onExpand={null}
  onSelect={null}
  onSort={null}
  rows={
    [
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
        ],
      },
    ]
  }
  summary={null}
  variant={null}
/>
`;

exports[`TableSkeleton Component should allow variations in table layout: className and variant 1`] = `
<Table
  ariaLabel={null}
  className="curiosity-skeleton-table lorem-ipsum-class"
  columnHeaders={
    [
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
    ]
  }
  componentClassNames={
    {
      "table": "curiosity-table",
      "td": "curiosity-table__td",
      "tdAction": "curiosity-table__td-action",
      "tdExpand": "curiosity-table__td-expand",
      "tdExpanded": "curiosity-table__td-expand-expanded",
      "tdExpandedContent": "curiosity-table__td-expand-content",
      "tdExpandedWrapper": "curiosity-table__td-expand-wrapper",
      "tdSelect": "curiosity-table__td-select",
      "th": "curiosity-table__th",
      "tr": "curiosity-table__tr",
      "trExpand": "curiosity-table__tr-expand",
      "trExpanded": "curiosity-table__tr-expand-expanded",
      "trExpandedContent": "curiosity-table__tr-expand-content",
    }
  }
  emptyTable={
    {
      "message": "Clear all filters and try again",
      "title": "No results found",
    }
  }
  isBorders={true}
  isHeader={false}
  isStriped={false}
  onExpand={null}
  onSelect={null}
  onSort={null}
  rows={
    [
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
        ],
      },
    ]
  }
  summary={null}
  variant="compact"
/>
`;

exports[`TableSkeleton Component should allow variations in table layout: column and row count  1`] = `
<Table
  ariaLabel={null}
  className="curiosity-skeleton-table "
  columnHeaders={
    [
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
    ]
  }
  componentClassNames={
    {
      "table": "curiosity-table",
      "td": "curiosity-table__td",
      "tdAction": "curiosity-table__td-action",
      "tdExpand": "curiosity-table__td-expand",
      "tdExpanded": "curiosity-table__td-expand-expanded",
      "tdExpandedContent": "curiosity-table__td-expand-content",
      "tdExpandedWrapper": "curiosity-table__td-expand-wrapper",
      "tdSelect": "curiosity-table__td-select",
      "th": "curiosity-table__th",
      "tr": "curiosity-table__tr",
      "trExpand": "curiosity-table__tr-expand",
      "trExpanded": "curiosity-table__tr-expand-expanded",
      "trExpandedContent": "curiosity-table__tr-expand-content",
    }
  }
  emptyTable={
    {
      "message": "Clear all filters and try again",
      "title": "No results found",
    }
  }
  isBorders={true}
  isHeader={false}
  isStriped={false}
  onExpand={null}
  onSelect={null}
  onSort={null}
  rows={
    [
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
        ],
      },
    ]
  }
  summary={null}
  variant={null}
/>
`;

exports[`TableSkeleton Component should allow variations in table layout: no table header and zero rows 1`] = `
<Table
  ariaLabel={null}
  className="curiosity-skeleton-table curiosity-skeleton-table__hidden-rows "
  columnHeaders={
    [
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
    ]
  }
  componentClassNames={
    {
      "table": "curiosity-table",
      "td": "curiosity-table__td",
      "tdAction": "curiosity-table__td-action",
      "tdExpand": "curiosity-table__td-expand",
      "tdExpanded": "curiosity-table__td-expand-expanded",
      "tdExpandedContent": "curiosity-table__td-expand-content",
      "tdExpandedWrapper": "curiosity-table__td-expand-wrapper",
      "tdSelect": "curiosity-table__td-select",
      "th": "curiosity-table__th",
      "tr": "curiosity-table__tr",
      "trExpand": "curiosity-table__tr-expand",
      "trExpanded": "curiosity-table__tr-expand-expanded",
      "trExpandedContent": "curiosity-table__tr-expand-content",
    }
  }
  emptyTable={
    {
      "message": "Clear all filters and try again",
      "title": "No results found",
    }
  }
  isBorders={true}
  isHeader={false}
  isStriped={false}
  onExpand={null}
  onSelect={null}
  onSort={null}
  rows={
    [
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
        ],
      },
    ]
  }
  summary={null}
  variant={null}
/>
`;

exports[`TableSkeleton Component should allow variations in table layout: table column widths 1`] = `
<Table
  ariaLabel={null}
  className="curiosity-skeleton-table "
  columnHeaders={
    [
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
        "width": 30,
      },
      {
        "content": <Skeleton
          size="md"
        />,
        "width": 15,
      },
      {
        "content": <Skeleton
          size="md"
        />,
        "width": 15,
      },
    ]
  }
  componentClassNames={
    {
      "table": "curiosity-table",
      "td": "curiosity-table__td",
      "tdAction": "curiosity-table__td-action",
      "tdExpand": "curiosity-table__td-expand",
      "tdExpanded": "curiosity-table__td-expand-expanded",
      "tdExpandedContent": "curiosity-table__td-expand-content",
      "tdExpandedWrapper": "curiosity-table__td-expand-wrapper",
      "tdSelect": "curiosity-table__td-select",
      "th": "curiosity-table__th",
      "tr": "curiosity-table__tr",
      "trExpand": "curiosity-table__tr-expand",
      "trExpanded": "curiosity-table__tr-expand-expanded",
      "trExpandedContent": "curiosity-table__tr-expand-content",
    }
  }
  emptyTable={
    {
      "message": "Clear all filters and try again",
      "title": "No results found",
    }
  }
  isBorders={true}
  isHeader={false}
  isStriped={false}
  onExpand={null}
  onSelect={null}
  onSort={null}
  rows={
    [
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
            "width": 30,
          },
          {
            "content": <Skeleton
              size="md"
            />,
            "width": 15,
          },
          {
            "content": <Skeleton
              size="md"
            />,
            "width": 15,
          },
        ],
      },
    ]
  }
  summary={null}
  variant={null}
/>
`;

exports[`TableSkeleton Component should allow variations in table layout: table header and zero rows 1`] = `
<Table
  ariaLabel={null}
  className="curiosity-skeleton-table curiosity-skeleton-table__hidden-rows "
  columnHeaders={
    [
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
      {
        "content": <Skeleton
          size="md"
        />,
      },
    ]
  }
  componentClassNames={
    {
      "table": "curiosity-table",
      "td": "curiosity-table__td",
      "tdAction": "curiosity-table__td-action",
      "tdExpand": "curiosity-table__td-expand",
      "tdExpanded": "curiosity-table__td-expand-expanded",
      "tdExpandedContent": "curiosity-table__td-expand-content",
      "tdExpandedWrapper": "curiosity-table__td-expand-wrapper",
      "tdSelect": "curiosity-table__td-select",
      "th": "curiosity-table__th",
      "tr": "curiosity-table__tr",
      "trExpand": "curiosity-table__tr-expand",
      "trExpanded": "curiosity-table__tr-expand-expanded",
      "trExpandedContent": "curiosity-table__tr-expand-content",
    }
  }
  emptyTable={
    {
      "message": "Clear all filters and try again",
      "title": "No results found",
    }
  }
  isBorders={true}
  isHeader={true}
  isStriped={false}
  onExpand={null}
  onSelect={null}
  onSort={null}
  rows={
    [
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
          {
            "content": <Skeleton
              size="md"
            />,
          },
        ],
      },
    ]
  }
  summary={null}
  variant={null}
/>
`;

exports[`TableSkeleton Component should render a non-connected component: non-connected 1`] = `
<Table
  ariaLabel={null}
  className="curiosity-skeleton-table "
  columnHeaders={
    [
      {
        "content": <Skeleton
          size="md"
        />,
      },
    ]
  }
  componentClassNames={
    {
      "table": "curiosity-table",
      "td": "curiosity-table__td",
      "tdAction": "curiosity-table__td-action",
      "tdExpand": "curiosity-table__td-expand",
      "tdExpanded": "curiosity-table__td-expand-expanded",
      "tdExpandedContent": "curiosity-table__td-expand-content",
      "tdExpandedWrapper": "curiosity-table__td-expand-wrapper",
      "tdSelect": "curiosity-table__td-select",
      "th": "curiosity-table__th",
      "tr": "curiosity-table__tr",
      "trExpand": "curiosity-table__tr-expand",
      "trExpanded": "curiosity-table__tr-expand-expanded",
      "trExpandedContent": "curiosity-table__tr-expand-content",
    }
  }
  emptyTable={
    {
      "message": "Clear all filters and try again",
      "title": "No results found",
    }
  }
  isBorders={true}
  isHeader={false}
  isStriped={false}
  onExpand={null}
  onSelect={null}
  onSort={null}
  rows={
    [
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
        ],
      },
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
        ],
      },
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
        ],
      },
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
        ],
      },
      {
        "cells": [
          {
            "content": <Skeleton
              size="md"
            />,
          },
        ],
      },
    ]
  }
  summary={null}
  variant={null}
/>
`;
